src/igmpproxy.c

```diff
@@ -140,18 +143,46 @@ int main( int ArgCn, char *ArgVc[] ) {
             break;
         }
 
+        // Open /dev/null before chrooting.
+        if (!NotAsDaemon) {
+            devnull = open("/dev/null", O_RDWR, 0);
+            if (devnull == -1)
+                my_log(LOG_ERR, 0, "unable to open /dev/null");
+        }
+
+        config = getCommonConfig();
+
+        if (strlen(config->user) != 0) {
+            pw = getpwnam(config->user);
+            if (pw == NULL)
+                my_log(LOG_ERR, 0, "unknown user %s", config->user);
+        }
+
+        if (strlen(config->chroot) != 0)
+            if (chroot(config->chroot) != 0 || chdir("/") != 0)
+                my_log(LOG_ERR, 0, "unable to chroot to %s",
+                  config->chroot);
+
+        if (strlen(config->user) != 0 && pw != NULL)
+            if (setgroups(1, &pw->pw_gid) != 0 ||
+              setresgid(pw->pw_gid, pw->pw_gid, pw->pw_gid) != 0 ||
+              setresuid(pw->pw_uid, pw->pw_uid, pw->pw_uid) != 0)
+                my_log(LOG_ERR, 0, "unable to drop privileges to %s",
+                  config->user);
```

Comment:
@pali, most builds in the CI run successfully. There are however a couple of build jobs that fail because the compiler (gcc-4.4 -> gcc-4.8) thinks a variable may be used uninitialized.
```igmpproxy.c: In function ‘main’:
igmpproxy.c:167:30: error: ‘pw’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
             if (setgroups(1, &pw->pw_gid) != 0 ||
                              ^
```
I do not understand why the compiler thinks this as the struct `pw` is checked before usage. I'm not sure that the code is at fault, and I'm thinking of selectively disabling this warning using something like below. What do you think, any ideas?
```
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
...
#pragma GCC diagnostic pop
```

Reply:
Because you forgot to initialize it, on line 85 you have just:
```c
struct passwd *pw;
```
Fix it properly instead of adding compiler warnings hacks.

Also on line 166 you are testing `config->user` member but you are not using it at all. So you should not test it. Just `pw` is enough.

And instead of counting length of string via `strlen(config->user)` you just need to know if is non-empty, which is idiomatically tested just by `config->user[0]`.

Reply:
Thanks!